File: [2023-08]/[5주차]/hjun-park/1916-최소비용구하기.java
import java.util.*;

public class _1916_최소비용_구하기 {
	static class Node implements Comparable<Node> {
		int node, distance;

		public Node(int node, int distance) {
			this.node = node;
			this.distance = distance;
		}

		@Override
		public int compareTo(Node other) {
			return Integer.compare(this.distance, other.distance);
		}
	}

	static int N, M;
	static List<int[]>[] graph;
	static int[] distance;
	static final int INF = (int) 1e9;

	public static void main(String[] args) {
		Scanner scanner = new Scanner(System.in);
		N = scanner.nextInt();
		M = scanner.nextInt();
		graph = new ArrayList[N + 1];
		distance = new int[N + 1];
		Arrays.fill(distance, INF);

		for (int i = 1; i <= N; i++) {
			graph[i] = new ArrayList<>();
		}

		for (int i = 0; i < M; i++) {
			int a = scanner.nextInt();
			int b = scanner.nextInt();
			int c = scanner.nextInt();
			graph[a].add(new int[]{b, c});
		}

		int start = scanner.nextInt();
		int end = scanner.nextInt();

		dijkstra(start);

		System.out.println(distance[end]);
	}

	static void dijkstra(int start) {
		PriorityQueue<Node> hq = new PriorityQueue<>();
		distance[start] = 0;
		hq.offer(new Node(start, 0));

		while (!hq.isEmpty()) {
			Node current = hq.poll();
			int now = current.node;
			int dist = current.distance;

			if (dist > distance[now]) {
				continue;
			}

			for (int[] next : graph[now]) {
				int nextNode = next[0];
				int cost = dist + next[1];

				if (cost < distance[nextNode]) {
					distance[nextNode] = cost;
					hq.offer(new Node(nextNode, cost));
				}
			}
		}
	}
}

File: [2023-08]/[5주차]/hjun-park/1916-최소비용구하기.py
'''
0. v+1 만큼 빈 graph 생성, 거리 distance 생성 후 모두 INF 초기화
1. 값 입력받아 인접리스트 생성
2. start 지점만 거리 0으로 설정
3. 다익스트라 시작
 3-1. 최초 hq에 담음 (거리, 노드)
 3-2. while hq 시작
    1) 거리와 노드를 꺼냄
    2) 꺼낸 거리가 distance 보다 더 짧다면 이미 갱신된 것이므로 return
    3) 그게 아니라면 graph[node] 순회 시작
      i) cost는 현재 꺼낸 거리 + 이전 노드 거리
      ii) 만약 cost가 기존 노드보다 짧다면 distance 갱신 후 heapq에 넣어준다.
        ==> 넣어주는 이유는 거리가 짧으니까
'''
import heapq

N = int(input().rstrip())  # vertex
M = int(input().rstrip())  # edge

INF = int(1e9)
graph = [[] for _ in range(N + 1)]
distance = [INF for _ in range(N + 1)]

for _ in range(M):
    a, b, c = map(int, input().split())
    graph[a].append([b, c])

start, end = map(int, input().split())


def dijkstra():
    hq = []

    distance[start] = 0
    heapq.heappush(hq, [0, start])

    while hq:
        dist, now = heapq.heappop(hq)

        if dist > distance[now]:
            continue

        # [node, distance]
        for nxt in graph[now]:
            cost = dist + nxt[1]

            if cost < distance[nxt[0]]:
                distance[nxt[0]] = cost
                heapq.heappush(hq, [cost, nxt[0]])


dijkstra()

print(distance[end])
